pkg/reconciler/v1alpha1/trigger/trigger.go

```diff
@@ -0,0 +1,687 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// controllerAgentName is the string used by this controller to identify
+	// itself when creating events.
+	controllerAgentName = "trigger-controller"
+
+	finalizerName = controllerAgentName
+
+	// Name of the corev1.Events emitted from the reconciliation process.
+	triggerReconciled         = "TriggerReconciled"
+	triggerReconcileFailed    = "TriggerReconcileFailed"
+	triggerUpdateStatusFailed = "TriggerUpdateStatusFailed"
+	subscriptionDeleteFailed  = "SubscriptionDeleteFailed"
+	subscriptionCreateFailed  = "SubscriptionCreateFailed"
+)
+
+var dummyValue struct{}
+
+type reconciler struct {
+	client        client.Client
+	restConfig    *rest.Config
+	dynamicClient dynamic.Interface
+	recorder      record.EventRecorder
+
+	triggersLock sync.RWMutex
+	// Contains the triggers that correspond to a particular broker.
+	// We use this to reconcile only the triggers that correspond to a certain broker.
+	// brokerNamespacedName -> triggerReconcileRequest -> dummy struct
+	triggers map[types.NamespacedName]map[reconcile.Request]struct{}
+
+	logger *zap.Logger
+}
+
+// Verify the struct implements reconcile.Reconciler.
+var _ reconcile.Reconciler = &reconciler{}
+
+// ProvideController returns a function that returns a Trigger controller.
+func ProvideController(mgr manager.Manager, logger *zap.Logger) (controller.Controller, error) {
+	// Setup a new controller to Reconcile Triggers.
+	r := &reconciler{
+		recorder: mgr.GetRecorder(controllerAgentName),
+		logger:   logger,
+		triggers: make(map[types.NamespacedName]map[reconcile.Request]struct{}),
+	}
+	c, err := controller.New(controllerAgentName, mgr, controller.Options{
+		Reconciler: r,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Watch Triggers.
+	if err = c.Watch(&source.Kind{Type: &v1alpha1.Trigger{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+
+	// Watch all the resources that the Trigger reconciles.
+	for _, t := range []runtime.Object{&corev1.Service{}, &istiov1alpha3.VirtualService{}, &v1alpha1.Subscription{}} {
+		err = c.Watch(&source.Kind{Type: t}, &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.Trigger{}, IsController: true})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Watch for Broker changes.
+	if err = c.Watch(&source.Kind{Type: &v1alpha1.Broker{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: &mapAllTriggers{r}}); err != nil {
+		return nil, err
+	}
+
+	// TODO reconcile after a change to the subscriber. I'm not sure how this is possible, but we should do it if we
+	// can find a way.
+
+	return c, nil
+}
+
+// mapAllTriggers maps Broker change notifications to Trigger reconcileRequests.
+type mapAllTriggers struct {
+	r *reconciler
+}
+
+func (m *mapAllTriggers) Map(o handler.MapObject) []reconcile.Request {
```

Comment:
Switched to using a `List` request, but I don't think I can use `MatchingField`, because that turns into a `FieldSelector` under the hood and those are only supported on certain fields: ["Supported field selectors vary by Kubernetes resource type...Using unsupported field selectors produces an error."](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#supported-fields). And ["CRDs still only support namespace and name for field selectors"](https://github.com/kubernetes/kubernetes/issues/51046#issuecomment-434040437).